store.go

```diff
@@ -26,39 +26,59 @@ type IpldBlockstore interface {
 	Put(block.Block) error
 }
 
+// IpldBlockstoreViewer is a trait that enables zero-copy access to blocks in
+// a blockstore.
+type IpldBlockstoreViewer interface {
+	// View provides zero-copy access to blocks in a blockstore. The callback
+	// function will be invoked with the value for the key. The user MUST not
+	// modify the byte array, as it could be memory-mapped.
+	View(cid.Cid, func([]byte) error) error
+}
+
 // BasicIpldStore wraps and IpldBlockstore and implements the IpldStore interface.
 type BasicIpldStore struct {
 	Blocks IpldBlockstore
-	Atlas  *atlas.Atlas
+	Viewer IpldBlockstoreViewer
+
+	Atlas *atlas.Atlas
 }
 
 var _ IpldStore = &BasicIpldStore{}
 
 // NewCborStore returns an IpldStore implementation backed by the provided IpldBlockstore.
 func NewCborStore(bs IpldBlockstore) *BasicIpldStore {
-	return &BasicIpldStore{Blocks: bs}
+	viewer, _ := bs.(IpldBlockstoreViewer)
+	return &BasicIpldStore{Blocks: bs, Viewer: viewer}
 }
 
 // Get reads and unmarshals the content at `c` into `out`.
 func (s *BasicIpldStore) Get(ctx context.Context, c cid.Cid, out interface{}) error {
-	blk, err := s.Blocks.Get(c)
-	if err != nil {
-		return err
-	}
+	f := func(b []byte) error {
```

Comment:
can we avoid the extra function call on each get?

We can do the direct atlas decode on the get path, and declare the same thing as a private method rather than an anonymous function that has to be re-allocated each `Get` that we use in the View case.

Reply:
Pretty sure this would get inlined, but let me check.

Reply:
No inlining happens, and Go benchmark shows that there's an extra alloc. I pushed 4c9b009 to incur in the (inevitable) callback alloc only in the zero-copy path.